src/Tools/dotnet-counters/CounterMonitor.cs

```diff
@@ -373,34 +373,32 @@ private async Task<int> Start()
 
             monitorTask.Start();
 
-            while(!shouldExit.WaitOne(250))
-            {
-                while (true)
+            
+            while (!shouldExit.WaitOne(250) && !(!Console.IsInputRedirected && Console.KeyAvailable)) { }
+            while (!shouldExit.WaitOne(0)) {
+                if (shouldExit.WaitOne(0))
+                {
+                    StopMonitor();
+                    return ReturnCode.Ok;
+                }
+                if (!Console.IsInputRedirected && Console.KeyAvailable)
                 {
-                    if (shouldExit.WaitOne(250))
+                    ConsoleKey cmd = Console.ReadKey(true).Key;
+                    if (cmd == ConsoleKey.Q)
                     {
                         StopMonitor();
                         return ReturnCode.Ok;
                     }
-                    if (Console.KeyAvailable)
+                    else if (cmd == ConsoleKey.P)
                     {
-                        break;
+                        pauseCmdSet = true;
+                    }
+                    else if (cmd == ConsoleKey.R)
+                    {
+                        pauseCmdSet = false;
                     }
                 }
```

Comment:
I think what you want is:

```csharp
			while (true)
            {
                if (shouldExit.WaitOne(250))
                {
                    StopMonitor();
                    break;
                }
				else if (!Console.IsInputRedirected && Console.KeyAvailable)
                {
                    ConsoleKey cmd = Console.ReadKey(true).Key;
                    if (cmd == ConsoleKey.Q)
                    {
                        StopMonitor();
                        break;
                    }
                    else if (cmd == ConsoleKey.P)
                    {
                        pauseCmdSet = true;
                    }
                    else if (cmd == ConsoleKey.R)
                    {
                        pauseCmdSet = false;
                    }
                }
			}
```

or

```csharp
			while (!shouldExit.WaitOne(250))
            {
                if (!Console.IsInputRedirected && Console.KeyAvailable)
                {
                    ConsoleKey cmd = Console.ReadKey(true).Key;
                    if (cmd == ConsoleKey.Q)
                    {
                        break;
                    }
                    else if (cmd == ConsoleKey.P)
                    {
                        pauseCmdSet = true;
                    }
                    else if (cmd == ConsoleKey.R)
                    {
                        pauseCmdSet = false;
                    }
                }
			}

			StopMonitor();
```

As written, there are two loops, but you only need the second one.